### PR TITLE
linalg: Moore-Penrose pseudo-inverse (`pinv`)

### PR DESCRIPTION
**Pull request by perazz**
Compute pseudo-inverse of a square or rectangular matrix using singular value decomposition (SVD). 
Array $A$, real or complex, has rank `n==2`. 

The pseudo-inverse $A^{+}$ is a generalization of the matrix inverse and satisfies the following properties:
- $A \cdot A^{+} \cdot A = A$
- $A^{+} \cdot A \cdot A^{+} = A^{+}$
- $(A \cdot A^{+})^H = A \cdot A^{+}$
- $(A^{+} \cdot A)^H = A^{+} \cdot A$

The user may specify an eigenvalue cutoff relative tolerance, `rtol`. 

### Proposed interface

- `ap = pinv(a [, rtol] [, err])`: function interface
- `call pseudoinvert(a, ap [, rtol] [, err])`: subroutine interface (preallocated `ap`)
- `.pinv.a`: operator interface

### Key facts
- The implementation is made to match that of `inv`, `invert` and `.inv.`
- [Fortran Discourse](https://fortran-lang.discourse.group/t/what-is-the-best-pseudo-inverse-moore-penrose-inverse-subroutine/3936) thread 
- Use SVD for performance: https://github.com/scipy/scipy/issues/12196
- if `rtol` is not provided, just use $\text{max}(m,n)\cdot\varepsilon$

### Progress
- [x] base implementation
- [x] tests
- [x] documentation  
- [x] submodule 
- [x] examples

### Prior art
- Numpy: `linalg.pinv(a, rcond=None, hermitian=False, *, rtol=<no value>)`
- Scipy: `pinv(a, *, atol=None, rtol=None, return_rank=False, check_finite=True)`
- Matlab: `B = pinv(a [,tol])`

cc: @jvdp1 @jalvesz @Beliavsky  @CRquantum @fortran-lang/stdlib

### CODE REVIEW

File: src/stdlib_linalg_pinv.fypp [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+         ! Get pseudo-inverse: A_pinv = V * (diag(1/s) * U^H) = V * (U * diag(1/s))^H
+         
+         ! 1) compute (U * diag(1/s)) in-place
+         forall (i=1:m,j=1:k) u(i,j) = s(j)*u(i,j)
+            
+         ! 2) commutate matmul: A_pinv = V^H * (U * diag(1/s))^H = ((U * diag(1/s)) * V^H)^H. 
+         !    This avoids one matrix transpose
+         #:if rt.startswith('complex')
+         pinva = conjg(transpose(matmul(u,vt)))
```
Comment (jalvesz) [inline src/stdlib_linalg_pinv.fypp:80]: These operations could be handled with `gemm` instead for max performance ?

**Review by jalvesz — CHANGES_REQUESTED** [review]

File: doc/specs/stdlib_linalg.md [hunk 2]
```diff
@@ -1459,6 +1459,142 @@ If `err` is not present, exceptions trigger an `error stop`.
...
+where \( \epsilon \) is the machine precision for the element type of `a`. It is an `intent(in)` argument.
+
+`err` (optional): Shall be a `type(linalg_state_type)` value. It is an `intent(out)` argument.
+
+### Return value
+
+Returns an array value of the same type, kind, and rank as `a` with shape `[n, m]`, that contains the pseudo-inverse matrix \( A^{+} \).
+
+Raises `LINALG_ERROR` if the underlying SVD did not converge.
+Raises `LINALG_VALUE_ERROR` if `pinva` and `a` have degenerate or incompatible sizes.
```
[the remark this answers is not in the document]
Comment (perazz) [reply doc/specs/stdlib_linalg.md:1501]:
```suggestion
Raises `LINALG_VALUE_ERROR` if `a` (or `pinva` if applicable) have degenerate or incompatible sizes.
```

File: src/stdlib_linalg_pinv.fypp [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+         call svd(a,s,u,vt,overwrite_a=.false.,full_matrices=.false.,err=err0)
+         if (err0%error()) then 
+            err0 = linalg_state_type(this,LINALG_ERROR,'svd failure -',err0%message)
+            call linalg_error_handling(err0,err)
+            return
+         endif
+         
+         !> Discard singular values
+         cutoff = tolerance*maxval(s)
+         s = merge(1/s,0.0_${rk}$,s>cutoff)
```
[the remark this answers is not in the document]
Comment (perazz) [reply src/stdlib_linalg_pinv.fypp:70]:
```suggestion
         s = merge(1.0_${rk}$/s,0.0_${rk}$,s>cutoff)
```

File: src/stdlib_linalg_pinv.fypp [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+         ! Get pseudo-inverse: A_pinv = V * (diag(1/s) * U^H) = V * (U * diag(1/s))^H
+         
+         ! 1) compute (U * diag(1/s)) in-place
+         forall (i=1:m,j=1:k) u(i,j) = s(j)*u(i,j)
```
[the remark this answers is not in the document]
Comment (perazz) [reply src/stdlib_linalg_pinv.fypp:75]:
```suggestion
         do concurrent (i=1:m,j=1:k); u(i,j) = s(j)*u(i,j); end do
```

File: src/stdlib_linalg_pinv.fypp (hunk 1 above)
Comment (perazz) [reply src/stdlib_linalg_pinv.fypp]: [c0f0e10](https://github.com/fortran-lang/stdlib/pull/899/commits/c0f0e10e31a7c0a5933a0e2e0c9d06dad5b22ffc)

**Review by jalvesz — APPROVED** [review]

File: doc/specs/stdlib_linalg.md [hunk 5]
```diff
@@ -1459,6 +1459,142 @@ If `err` is not present, exceptions trigger an `error stop`.
...
+- \( (A \cdot A^{+})^T = A \cdot A^{+} \)
+- \( (A^{+} \cdot A)^T = A^{+} \cdot A \)
+
+The computation is based on singular value decomposition (SVD). Singular values below a relative 
+tolerance threshold \( \text{rtol} \cdot \sigma_{\max} \), where \( \sigma_{\max} \) is the largest 
+singular value, are treated as zero.
+
+### Syntax
+
+`b ` [[stdlib_linalg(module):pinv(interface)]] `(a, [, rtol, err])`
```
Comment (jvdp1) [inline doc/specs/stdlib_linalg.md:1483]:
```suggestion
`b =` [[stdlib_linalg(module):pinv(interface)]] `(a, [, rtol, err])`
```

File: doc/specs/stdlib_linalg.md (hunk 2 above)
Comment (jvdp1) [inline doc/specs/stdlib_linalg.md:1501]: `pinva` is not mentioned in the description of `pinv`. So this sounds strange to read it here.

File: doc/specs/stdlib_linalg.md [hunk 6]
```diff
@@ -1459,6 +1459,142 @@ If `err` is not present, exceptions trigger an `error stop`.
 {!example/linalg/example_inverse_function.f90!}
 ```
 
+## `pinv` - Moore-Penrose pseudo-inverse of a matrix.
```
Comment (jvdp1) [inline doc/specs/stdlib_linalg.md:1462]:
```suggestion
## `pinv` - Moore-Penrose pseudo-inverse of a matrix
```

**Review by jvdp1 — APPROVED** [review]
Thank you @perazz . It looks really good to me!

File: doc/specs/stdlib_linalg.md (hunk 2 above)
Comment (perazz) [reply doc/specs/stdlib_linalg.md]: [9dc5cc9](https://github.com/fortran-lang/stdlib/pull/899/commits/9dc5cc960293e7e20ac76c6f63c6f9c97f99fe30) Thanks, that was copied too quickly from `pseudoinvert`, fixed now.

**Comment by perazz** [comment]
Thank you for the reviews and the approvals @jalvesz @jvdp1, I believe all fixes are addressed, so I will merge this.

**Pull request merged**